### PR TITLE
Feat(stableswap): withdraw/provide with fee

### PR DESCRIPTION
**Pull request by epanchee**

### CODE REVIEW

File: contracts/pair_stable/src/contract.rs [hunk 1]
```diff
@@ -591,7 +626,82 @@ pub fn withdraw_liquidity(
...
+            return Err(StdError::generic_err(format!(
+                "Not enough LP tokens. You need {} LP tokens.",
+                need_lp_tokens
+            ))
+            .into());
+        } else if need_lp_tokens.is_zero() {
+            return Err(
+                StdError::generic_err("Failed to calculate necessary LP tokens amount").into(),
+            );
+        }
```
Comment (ShadoySV) [inline contracts/pair_stable/src/contract.rs:699]: This block won't be ever executed due to line 688. The check should be done before adding one at line 688

**Review by ShadoySV — CHANGES_REQUESTED** [review]

File: contracts/pair_stable/src/contract.rs (hunk 1 above)
Comment (epanchee) [reply contracts/pair_stable/src/contract.rs:699]: That's true. Thanks

File: contracts/pair_stable/src/contract.rs [hunk 2]
```diff
@@ -454,8 +454,40 @@ pub fn provide_liquidity(
             return Err(ContractError::LiquidityAmountTooSmall {});
         }
 
+        // Get fee info from the factory
+        let fee_info = query_fee_info(
+            &deps.querier,
+            &config.factory_addr,
+            config.pair_info.pair_type.clone(),
+        )
+        .unwrap_or_default(); // There may no fee exist thus 0 is a default fee.
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/pair_stable/src/contract.rs:463]: Yep, will fix

File: contracts/pair_stable/src/contract.rs [hunk 3]
```diff
@@ -454,8 +454,40 @@ pub fn provide_liquidity(
...
+
+        let fee = fee_info.total_fee_rate / Uint128::from(2u8);
```
Comment (P-Yevhenii) [inline contracts/pair_stable/src/contract.rs:465]: Let's try do not use magic numbers, instead use constant if it possible. Or maybe add some comment above what they are means, math formula or additional description why we are use '2'.

File: contracts/pair_stable/src/contract.rs [hunk 4]
```diff
@@ -591,7 +626,82 @@ pub fn withdraw_liquidity(
...
+        let d1 = compute_d(amp, new_balances[0].u128(), new_balances[1].u128()).unwrap();
+
+        // Get fee info from the factory
+        let fee_info = query_fee_info(
+            &deps.querier,
+            &config.factory_addr,
+            config.pair_info.pair_type.clone(),
+        )?;
+
+        let fee = fee_info.total_fee_rate / Uint128::from(2u8);
```
Comment (P-Yevhenii) [inline contracts/pair_stable/src/contract.rs:668]: Also here

File: contracts/pair_stable_bluna/src/contract.rs [hunk 5]
```diff
@@ -536,8 +538,40 @@ pub fn provide_liquidity(
             return Err(ContractError::LiquidityAmountTooSmall {});
         }
 
+        // Get fee info from the factory
+        let fee_info = query_fee_info(
+            &deps.querier,
+            &config.factory_addr,
+            config.pair_info.pair_type.clone(),
+        )
+        .unwrap_or_default(); // There may no fee exist thus 0 is a default fee.
```
Comment (P-Yevhenii) [inline contracts/pair_stable_bluna/src/contract.rs:547]: As @ShadoySV said before, let's return error

File: contracts/pair_stable_bluna/src/contract.rs [hunk 6]
```diff
@@ -536,8 +538,40 @@ pub fn provide_liquidity(
...
+
+        let fee = fee_info.total_fee_rate / Uint128::from(2u8);
```
Comment (P-Yevhenii) [inline contracts/pair_stable_bluna/src/contract.rs:549]: Do not use magic numbers, please. If it possible use constant or add some comment

File: contracts/pair_stable/src/contract.rs (hunk 1 above)
Comment (P-Yevhenii) [reply contracts/pair_stable/src/contract.rs:699]: Need to fix it for pair_stable_bluna too

**Comment by ShadoySV** [comment]
The changes were moved into #281

**Pull request closed without merging**